examples/voila/app-with-auth0/voila.py

```diff
@@ -0,0 +1,12 @@
+import nbformat
+
+
+def hook_function(req, notebook, cwd):
+   headers = req.request.headers
+   notebook.metadata.user = headers.get('X-Auth-Name', 'Anonymous')
+   with open("app_updated.ipynb", 'w') as f:
```

Comment:
The code works:
<img width="650" alt="Screenshot 2024-05-14 at 7 05 27 PM" src="https://github.com/ploomber/doc/assets/29096135/fbd826e0-9954-461b-b4fd-0211851527a1">

But I'm a bit unsure of this saving notebook part. Initially I wanted to inject the username as metadata to the notebook and read inside the Voila app. but the metadata is not updated unless it is saved. I have used `nbformat` to read the file and was trying out [ipylab](https://github.com/jtpio/ipylab/blob/main/examples/commands.ipynb) but was having trouble forming the exact json command.
Please suggest if there is any better way to read the updated metadata within the notebook. I fnot maybe i can change this snippet to just save a json `{username': <X-Auth-Name value>}` and read this json file inside the Voila app. @edublancas

Reply:
is there any particular reason why you want to store it in the metadata?

reading the headers from the request as you've done in other examples is fine

Reply:
I was referring to the [Voila docs](https://voila.readthedocs.io/en/stable/customize.html#accessing-the-tornado-request-prelaunch-hook) and I think this is the way of reading headers in Voila?

Reply:
maybe I'm missing something but can't you do this?

```python
headers = req.request.headers
name = headers.get('X-Auth-Name', 'Anonymous')
print(name)
```

would this make the name visible?

Reply:
No this wont print in the notebook. this is just the Voila configuration file. When we create a `voila.py` file in the root directory Voila will start from this script and use the hook function. Adding print in this file will only print in the terminal. We need to somehow send this value to the notebook that we are running Voila with. 

I think for now I'll just put this value in a text/json file and have the notebook read from that file. @edublancas

Reply:
ok, I saw this is a voila.py file not the notebook file that contains the app. why do we need this?

what happens if you don't use this config file and put this directly into the notebook?

```python
headers = req.request.headers
name = headers.get('X-Auth-Name', 'Anonymous')
print(name)
```

Reply:
but I'm not sure how the notebook will have access to the `req` object. It's a `VoilaTornadoHandler`. I'll check

Reply:
As far as I understand from the docs I think the request is accessible through a prelaunch-hook function only. I couldn't find an example where the request object has been accessed directly in the notebook

Reply:
This is from the docs:

> In certain custom setups when you need to access the tornado request object in order to check for authentication cookies, access details about the request headers, or modify the notebook before rendering. You can leverage the prelaunch-hook, which lets you inject a function to inspect the notebook and the request prior to executing them.

Reply:
ok, if you cannot find a way, it's ok to store it in a file as you suggested

Reply:
[JSON storage](https://github.com/ploomber/doc/pull/224/commits/6e9f5bc96c49b43d9dcffae3e357136e8a0e963d)

Reply:
I just noticed something: what happens if two users are logged in?

say user A logs in, and then user B logs in. if user A refreshes the browser, it'll read the latest file and think it's user B, right?

I'm unsure why you discarded modifying the notebook using nbformat, that sounds like a good approach. if you're unable to modify the metadata, you can just inject a new code cell that declares a new variable like:

```
user = "some-user"
```

so then they user can refer to it in their code, right?

Reply:
I discarded the nbformat approach because sometimes when error was occurring it would clear the entire notebook. But yes I can try injecting a code cell. i didnt think of this approach

Reply:
Also, will this case arise where two users login to the same application? since both users will have different auth0 setup?

Reply:
[Fixed](https://github.com/ploomber/doc/pull/224/commits/0ec99ed3880f1bc55e62b23471ac519bb8f50f8b)

Reply:
ok, this solution makes sense. did you test it?

> Also, will this case arise where two users login to the same application? since both users will have different auth0 setup?

I'm not following this. which case? and what do you mean by auth0 setup?

Reply:
Yes I tested (latest result)
<img width="700" alt="Screenshot 2024-05-17 at 12 21 05 AM" src="https://github.com/ploomber/doc/assets/29096135/7b0245dc-2ba5-4d22-9149-4623ea6ecf8f">

> I'm not following this. which case? and what do you mean by auth0 setup?

I mean the case where you mentioned first user A will login and then user B. Ideally user A and user B will have their own auth0 credentials and callback URLs so will they be able to login to the same application?